### ISSUE TITLE
[MASSEMBLY-578] lineEnding parameter makes assembly ignore empty directories

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-assembly-plugin#730]
**[Elie Prudhomme](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=elie)** opened **[MASSEMBLY-578](https://issues.apache.org/jira/browse/MASSEMBLY-578?redirect=false)** and commented

Empty directories are ignored if the lineEnding parameter is used in a fileset.
Created archive will contain empty directories if I either remove the lineEnding parameter OR remove any text file from the directory structure.

structure:

```
pom.xml
src/main/assembly/assembly.xml
src/main/include/a/test.txt
                /b/
```

pom.xml

```
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>test</groupId>
    <artifactId>test</artifactId>
    <version>0-SNAPSHOT</version>
    <packaging>pom</packaging>
    <name>test</name>

    <build>
        <plugins>
            <plugin>
                <artifactId>maven-assembly-plugin</artifactId>
                <configuration>
                    <appendAssemblyId>false</appendAssemblyId>
                    <attach>false</attach>
                    <descriptors>
                        <descriptor>src/main/assembly/assembly.xml</descriptor>
                    </descriptors>
                </configuration>
            </plugin>
        </plugins>
    </build>
</project>
```

assembly.xml

```
<assembly
        xmlns="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.2"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.2 http://maven.apache.org/xsd/assembly-1.1.2.xsd">
    <id>distribution</id>

    <formats>
        <format>tar.gz</format>
    </formats>

    <fileSets>
        <fileSet>
            <directory>src/main/include</directory>
            <outputDirectory>/</outputDirectory>
            <lineEnding>unix</lineEnding>
        </fileSet>
    </fileSets>
</assembly>
```

**test**

```
mvn clean assembly:single

set MAVEN_OPTS="-Xmx256m"
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] Building test
[INFO]    task-segment: [clean, assembly:single]
[INFO] ------------------------------------------------------------------------
[INFO] [clean:clean {execution: default-clean}]
[INFO] Deleting directory C:\ATM\emptybug\trunk\target
[INFO] [assembly:single {execution: default-cli}]
[INFO] Reading assembly descriptor: src/main/assembly/assembly.xml
[INFO] Building tar : C:\ATM\emptybug\trunk\target\test-0-SNAPSHOT.tar.gz
[WARNING] Assembly file: C:\ATM\emptybug\trunk\target\test-0-SNAPSHOT.tar.gz is not a regular file (
it may be a directory). It cannot be attached to the project build for installation or deployment.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESSFUL
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 1 second
[INFO] Finished at: Thu Oct 06 09:06:23 EDT 2011
[INFO] Final Memory: 8M/162M
[INFO] ------------------------------------------------------------------------
bsdtar ztvf target\test-0-SNAPSHOT.tar.gz
drwxr-xr-x  0 0      0           0 Oct 06 09:06 test-0-SNAPSHOT/
drwxr-xr-x  0 0      0           0 Oct 06 09:06 test-0-SNAPSHOT/a/
-rw-r--r--  0 0      0           0 Oct 06 09:06 test-0-SNAPSHOT/a/test.txt
```

**without lineEnding**

---

**Affects:** 2.2-beta-2

**Attachments:**
- [assembly.xml](https://issues.apache.org/jira/secure/attachment/12714052/assembly.xml) (_631 bytes_)
- [pom.xml](https://issues.apache.org/jira/secure/attachment/12714359/pom.xml) (_928 bytes_)

**Issue Links:**
- [MASSEMBLY-108](https://issues.apache.org/jira/browse/MASSEMBLY-108) Assembly Plugin Implicitly Excludes Empty Directory

1 votes, 2 watchers

**Comment by jira-importer** [comment]
**[Elie Prudhomme](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=elie)** commented

**without** lineEnding parameter, empty directories are preserved
assembly.xml

```
<assembly
        xmlns="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.2"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.2 http://maven.apache.org/xsd/assembly-1.1.2.xsd">
    <id>distribution</id>

    <formats>
        <format>tar.gz</format>
    </formats>

    <fileSets>
        <fileSet>
            <directory>src/main/include</directory>
            <outputDirectory>/</outputDirectory>
        </fileSet>
    </fileSets>
</assembly>
```

```
mvn clean assembly:single
set MAVEN_OPTS="-Xmx256m"
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] Building test
[INFO]    task-segment: [clean, assembly:single]
[INFO] ------------------------------------------------------------------------
[INFO] [clean:clean {execution: default-clean}]
[INFO] Deleting directory C:\ATM\emptybug\trunk\target
[INFO] [assembly:single {execution: default-cli}]
[INFO] Reading assembly descriptor: src/main/assembly/assembly.xml
[INFO] Building tar : C:\ATM\emptybug\trunk\target\test-0-SNAPSHOT.tar.gz
[WARNING] Assembly file: C:\ATM\emptybug\trunk\target\test-0-SNAPSHOT.tar.gz is not a regular file (
it may be a directory). It cannot be attached to the project build for installation or deployment.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESSFUL
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 1 second
[INFO] Finished at: Thu Oct 06 09:18:29 EDT 2011
[INFO] Final Memory: 9M/164M
[INFO] ------------------------------------------------------------------------
bsdtar ztvf target\test-0-SNAPSHOT.tar.gz
drwxr-xr-x  0 0      0           0 Oct 06 08:44 test-0-SNAPSHOT/
drwxr-xr-x  0 0      0           0 Oct 06 08:50 test-0-SNAPSHOT/a/
drwxr-xr-x  0 0      0           0 Oct 06 08:51 test-0-SNAPSHOT/b/
-rw-r--r--  0 0      0           0 Oct 06 08:50 test-0-SNAPSHOT/a/test.txt
```

**Comment by jira-importer** [comment]
**[Elie Prudhomme](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=elie)** commented

A workaround is to specify the fileSet twice, once excluding all files that will get all directories including empty ones and another including everything and specifying the lineEnding parameter.
The resulting archive will contain duplicates with 2.2-beta-2, version 2.2-beta-3 will produce the result expected.

```
<assembly
        xmlns="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.2"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.2 http://maven.apache.org/xsd/assembly-1.1.2.xsd">
    <id>distribution</id>

    <formats>
        <format>tar.gz</format>
    </formats>

    <fileSets>
        <fileSet>
            <directory>src/main/include</directory>
            <outputDirectory>/</outputDirectory>
            <excludes>
                <exclude>**/*.*</exclude>
            </excludes>
        </fileSet>
        <fileSet>
            <directory>src/main/include</directory>
            <outputDirectory>/</outputDirectory>
	    <lineEnding>unix</lineEnding>
        </fileSet>
    </fileSets>
</assembly>
```

**Comment by jira-importer** [comment]
**[hildo](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hildo)** commented

I can confirm this occurs when using a format of "dir" as well.  I'm using Maven 3.0.3, and the 2.3 version of the assembly plugin.

Empty (sub) directories are retained if no lineEndings are specified, but lost if lineEndings are defined for the fileset.

**Comment by jira-importer** [comment]
**[Kristian Rosenvold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=krosenvold)** commented

This issue should be fixed with 2.5.
